File: stepic_algorithms/tmp/bellman_ford.py
from decimal import Decimal as d

class Edge(object):
    def __init__(self, source, target, weight):
        self.s = source
        self.t = target
        self.w = weight
    
class Graph(object):
    def __init__(self, n):
        self.has_cycle = False
        self.v = n
        self.e = set()

    def add_edge(self, e):
        self.e.add(e)

    def bf(self, s):
        if self.has_cycle: return

        self.dist = [d('Infinity') for i in range(0, self.v + 1)]
        self.dist[s] = 0
        
        for i in range(0, self.v):
            for edge in self.e:
                if self.dist[edge.s] > self.dist[edge.t] + edge.w:
                    self.dist[edge.s] = self.dist[edge.t] + edge.w

        for i in range(1):
            for edge in self.e:
                if self.dist[edge.s] > self.dist[edge.t] + edge.w:
                    self.has_cycle = True
                    return


if __name__ == '__main__':
    n, m = map(int, input().split())
    g = Graph(n)

    for i in range(0, m):
        s, t, w = map(int, input().split())
        g.add_edge(Edge(s, t, w))

    for i in range(1, g.v):
        if g.has_cycle: break
        g.bf(i)

    print(g.has_cycle)

File: stepic_algorithms/tmp/bellman_ford.cpp
#include <iostream>
#include <vector>

using namespace std;

const int INF = 100000;

struct Edge {
    int s, t, w;
};

bool bf(int source, int n, Edge e[], int m) {
    vector <int> d(n, INF);
    d[source] = 0;
    
    for (int i = 0; i < n-1; ++i) {
        for (int j = 0; j < m; ++j) {
            if (d[e[j].s] > d[e[j].t] + e[j].w) {
				d[e[j].s] = d[e[j].t] + e[j].w;
			}
        }
    }
    
    for (int i = 0; i < 1; ++i) {
        for (int j = 0; j < m; ++j) {
            if (d[e[j].s] > d[e[j].t] + e[j].w) {
	    		return true;
		    }
        }
    }

    return false;
}

int main() {
    int n,m,w,s,t;
    cin >> n >> m;

    Edge e[m];

    for (int i = 0; i < m; ++i) {
        cin >> s >> t >> w;
        e[i].s = s - 1;
        e[i].t = t - 1;
        e[i].w = w;
    }

    bool result;
    result = bf(1, n, e, m);
    if (result) cout << "True";
    else cout << "False";
    return 0;
}
